File: backend/app/routers/metadata.py
"""Metadata API route — item catalogs (weapons, skills, equipment, ammo)."""

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel

from app.database import get_session
from app.models.item import Weapon, Skill, Equipment, Ammunition

router = APIRouter(prefix="/api/metadata", tags=["metadata"])


class WeaponResponse(BaseModel):
    id: int
    name: str
    weapon_type: str | None = None
    burst: str | None = None
    damage: str | None = None
    saving: str | None = None
    saving_num: str | None = None
    properties: list[str] = []
    distance: dict | None = None
    wiki_url: str | None = None

    model_config = {"from_attributes": True}


class SkillResponse(BaseModel):
    id: int
    name: str
    wiki_url: str | None = None

    model_config = {"from_attributes": True}


class EquipmentResponse(BaseModel):
    id: int
    name: str
    wiki_url: str | None = None

    model_config = {"from_attributes": True}


class AmmunitionResponse(BaseModel):
    id: int
    name: str
    wiki_url: str | None = None

    model_config = {"from_attributes": True}


class MetadataResponse(BaseModel):
    weapons: list[WeaponResponse]
    skills: list[SkillResponse]
    equipment: list[EquipmentResponse]
    ammunitions: list[AmmunitionResponse]


@router.get("", response_model=MetadataResponse)
async def get_metadata(session: AsyncSession = Depends(get_session)):
    """Get all item catalogs (weapons, skills, equipment, ammunition)."""
    weapons = (await session.execute(select(Weapon).order_by(Weapon.name))).scalars().all()
    skills = (await session.execute(select(Skill).order_by(Skill.name))).scalars().all()
    equips = (await session.execute(select(Equipment).order_by(Equipment.name))).scalars().all()
    ammos = (await session.execute(select(Ammunition).order_by(Ammunition.name))).scalars().all()

    return MetadataResponse(
        weapons=[WeaponResponse.model_validate(w) for w in weapons],
        skills=[SkillResponse.model_validate(s) for s in skills],
        equipment=[EquipmentResponse.model_validate(e) for e in equips],
        ammunitions=[AmmunitionResponse.model_validate(a) for a in ammos],
    )

File: backend-ts/src/routes/metadata.ts
import { Hono } from 'hono';
import { asc } from 'drizzle-orm';
import { db } from '../db/client.js';
import { weapons, skills, equipment, ammunitions } from '../db/schema.js';

const router = new Hono();

router.get('/', async c => {
    const [w, s, e, a] = await Promise.all([
        db.select().from(weapons).orderBy(asc(weapons.name)),
        db.select().from(skills).orderBy(asc(skills.name)),
        db.select().from(equipment).orderBy(asc(equipment.name)),
        db.select().from(ammunitions).orderBy(asc(ammunitions.name)),
    ]);

    return c.json({
        weapons: w.map(r => ({
            id: r.id,
            name: r.name,
            weapon_type: r.weapon_type,
            burst: r.burst,
            damage: r.damage,
            saving: r.saving,
            saving_num: r.saving_num,
            properties: Array.isArray(r.properties) ? r.properties : [],
            distance: r.distance ?? null,
            wiki_url: r.wiki_url,
        })),
        skills: s.map(r => ({ id: r.id, name: r.name, wiki_url: r.wiki_url })),
        equipment: e.map(r => ({ id: r.id, name: r.name, wiki_url: r.wiki_url })),
        ammunitions: a.map(r => ({ id: r.id, name: r.name, wiki_url: r.wiki_url })),
    });
});

export default router;
